GetMessages/handler.ts

```diff
@@ -50,42 +67,110 @@ type IGetMessagesHandlerResponse =
  *
- * TODO: add full results and paging
  */
 type IGetMessagesHandler = (
-  fiscalCode: FiscalCode
+  fiscalCode: FiscalCode,
+  maybePageSize: O.Option<NonNegativeInteger>,
+  maybeEnrichResultData: O.Option<boolean>,
+  maybeMaximumId: O.Option<NonEmptyString>,
+  maybeMinimumId: O.Option<NonEmptyString>
 ) => Promise<IGetMessagesHandlerResponse>;
 
 /**
  * Handles requests for getting all message for a recipient.
  */
-export function GetMessagesHandler(
-  messageModel: MessageModel
-): IGetMessagesHandler {
-  return async fiscalCode => {
-    return pipe(
-      messageModel.findMessages(fiscalCode),
-      TE.map(flattenAsyncIterator),
-      TE.map(_ => filterAsyncIterator(_, isRight)),
-      TE.map(_ => mapAsyncIterator(_, e => e.right)),
-      TE.map(_ => filterAsyncIterator(_, RetrievedNotPendingMessage.is)),
-      TE.map(_ => mapAsyncIterator(_, retrievedMessageToPublic)),
-      TE.bimap(
-        failure => ResponseErrorQuery(failure.kind, failure),
-        ResponseJsonIterator
-      ),
-      TE.toUnion
-    )();
-  };
-}
+export const GetMessagesHandler = (
+  messageModel: MessageModel,
+  serviceModel: ServiceModel,
+  blobService: BlobService
+): IGetMessagesHandler => async (
+  fiscalCode,
+  maybePageSize,
+  maybeEnrichResultData,
+  maybeMaximumId,
+  maybeMinimumId
+) =>
+  pipe(
+    TE.Do,
+    TE.bind("pageSize", () =>
+      TE.of(O.getOrElse(() => defaultPageSize)(maybePageSize))
+    ),
+    TE.bind("shouldEnrichResultData", () =>
+      TE.of(O.getOrElse(() => false)(maybeEnrichResultData))
+    ),
+    TE.bind("maximumId", () => TE.of(O.toUndefined(maybeMaximumId))),
+    TE.bind("minimumId", () => TE.of(O.toUndefined(maybeMinimumId))),
+    TE.chain(({ pageSize, shouldEnrichResultData, maximumId, minimumId }) =>
+      pipe(
+        messageModel.findMessages(fiscalCode, pageSize, maximumId, minimumId),
+        TE.mapLeft(failure => ResponseErrorQuery(failure.kind, failure)),
+        TE.map(i => mapAsyncIterator(i, A.rights)),
+        TE.map(i =>
+          mapAsyncIterator(i, A.filter(RetrievedNotPendingMessage.is))
+        ),
+        TE.map(i => mapAsyncIterator(i, A.map(retrievedMessageToPublic))),
+        TE.chainW(i =>
+          // check whether we should enrich messages or not
+          pipe(
+            TE.fromPredicate(
+              () => shouldEnrichResultData === true,
+              () =>
+                // if no enrichment is requested we just wrap messages in a TE
+                mapAsyncIterator(
+                  i,
+                  // A.map(e => TE.of<Error, CreatedMessageWithoutContent>(e))
+                  A.map(async e =>
+                    E.right<Error, CreatedMessageWithoutContent>(e)
+                  )
+                )
+            )(i),
+            TE.map(j =>
+              mapAsyncIterator(
+                j,
+                enrichMessagesData(messageModel, serviceModel, blobService)
+              )
+            ),
+            TE.orElse(TE.of),
```

Comment:
I understand the need of this line, although it's a bit cryptic. Consider add a comment to explain the trick to our future selves :)

Reply:
Fixe in https://github.com/pagopa/io-functions-app/pull/203/commits/1cf1b28e1b0a804d1843054480a4029665ad1e53